### DISCUSSION TITLE
Concurrency vs Parallelism

### DISCUSSION THREAD

**Discussion by mcieaaka** [sidekiq/sidekiq#6119] [Q&A] [answered]
Before getting into the question, I am assuming everyone knows the difference between the two and here is an article just for context: https://workingwithruby.com/wwrt/concurrent_vs_parallel/

Now when it comes to sidekiq, let us assume I have 2 kinds of applications:
A) RAILS APP: MRI (CRuby) WITH GIL
B) RAILS APP: JRuby WITHOUT GIL

Now, both these applications are using  Sidekiq for scheduling and executing background jobs.
Now with respect to the following job I have a few doubts:
```
class DemoWorker
	include Sidekiq::Worker

	def perform
		@threads = []
		spawn_threads_and_perform_action
	end

	attr_accessor :threads

        # Looking at the method it is obvious that it will spawn only 4 new threads
	def spawn_threads_and_perform_action
		MyModel.limit(100).find_in_batches(batch_size: 25).each do |batch|
			threads << Thread.new do
				# Do something on batch
			end
		end

		threads.map(&:join)
	end
end
```

In all cases we know for a fact that the "threads" will be executed concurrently but assume I have 4 Cores in my CPU

1. In Application A, will the threads execute parallelly, why or why not ?
2. In Application B, will the threads execute parallelly, why or why not ?
3. Will both A and B execute Parallelly irrespective of the application's compiler, why or why not ?

**Comment by mperham** [comment, marked as answer]
In JRuby they will execute in parallel. In MRI they will execute concurrently.
